compiler/rustc_lint/src/default_could_be_derived.rs

```diff
@@ -0,0 +1,422 @@
+                                            && let Some(prev_field) = fields.get(i - 1)
+                                        {
+                                            removals.push(
+                                                prev_field
+                                                    .span
+                                                    .shrink_to_hi()
+                                                    .with_hi(field.span.hi()),
+                                            );
+                                        } else {
+                                            removals.push(field.span);
+                                        }
+                                    }
+                                }
+                                let mut sugg: Vec<(Span, String)> =
+                                    removals.into_iter().map(|sp| (sp, String::new())).collect();
+                                match tail {
+                                    hir::StructTailExpr::Base(_) => {}
+                                    hir::StructTailExpr::DefaultFields(_) => {}
+                                    hir::StructTailExpr::None(span) => {
+                                        sugg.push((span, ", ..".to_string()));
+                                    }
+                                }
+                                diag.multipart_suggestion_verbose(
+                                    "use the default values in the `impl` to avoid them diverging \
+                                     over time",
+                                    sugg,
+                                    Applicability::MachineApplicable,
+                                );
+                            },
+                        );
+                    }
+                }
+                _ => {}
+            }
+        }
+    }
+}
+
+fn check_path<'tcx>(
+    cx: &LateContext<'tcx>,
+    path: &hir::QPath<'_>,
+    hir_id: hir::HirId,
+    ty: Ty<'tcx>,
+) -> bool {
+    let Some(default_def_id) = cx.tcx.get_diagnostic_item(sym::Default) else {
+        return false;
+    };
+    let res = cx.qpath_res(&path, hir_id);
+    let Some(def_id) = res.opt_def_id() else { return false };
+    if cx.tcx.is_diagnostic_item(sym::default_fn, def_id) {
+        // We have `field: Default::default(),`. This is what the derive would do already.
+        return true;
+    }
+    // For every `Default` impl for this type (there should be a single one), we see if it
+    // has a "canonical" `DefId` for a fn call with no arguments, or a path. If it does, we
+    // check that `DefId` with the `DefId` of this field's value if it is also a call/path.
+    // If there's a match, it means that the contents of that type's `Default` impl are the
+    // same to what the user wrote on *their* `Default` impl for this field.
+    let mut equivalents = vec![];
+    cx.tcx.for_each_relevant_impl(default_def_id, ty, |impl_def_id| {
```

Comment:
This is certainly not the right way to look up a default impl for a given type. Use `Instance::try_resolve`.